compiler/noirc_evaluator/src/acir/call/intrinsics/slice_ops.rs

```diff
@@ -502,61 +501,57 @@ impl Context<'_> {
+        for (i, (current_value, _)) in flat_slice.iter().enumerate().take(result_size) {
             let current_index = self.acir_context.add_constant(i);
 
-            let value_current_index = &new_slice[i].borrow_var()?;
+            let shifted_index = self.acir_context.add_constant(i + popped_elements_size);
 
-            if (i + popped_elements_size) < slice_size {
-                let shifted_index = self.acir_context.add_constant(i + popped_elements_size);
-
-                let value_shifted_index =
-                    self.acir_context.read_from_memory(block_id, &shifted_index)?;
+            // Fetch the value from the initial slice
+            let value_shifted_index =
+                self.acir_context.read_from_memory(block_id, &shifted_index)?;
 
-                let use_shifted_value =
-                    self.acir_context.more_than_eq_var(current_index, flat_user_index, 64)?;
+            let use_shifted_value =
+                self.acir_context.more_than_eq_var(current_index, flat_user_index, 64)?;
 
-                let shifted_value_pred =
-                    self.acir_context.mul_var(value_shifted_index, use_shifted_value)?;
-                let not_pred = self.acir_context.sub_var(one, use_shifted_value)?;
-                let current_value_pred =
-                    self.acir_context.mul_var(not_pred, *value_current_index)?;
+            let shifted_value_pred =
+                self.acir_context.mul_var(value_shifted_index, use_shifted_value)?;
+            let not_pred = self.acir_context.sub_var(one, use_shifted_value)?;
+            let current_value_pred = self.acir_context.mul_var(not_pred, *current_value)?;
 
-                let new_value =
-                    self.acir_context.add_var(shifted_value_pred, current_value_pred)?;
+            let new_value = self.acir_context.add_var(shifted_value_pred, current_value_pred)?;
 
-                self.acir_context.write_to_memory(result_block_id, &current_index, &new_value)?;
-            };
+            self.acir_context.write_to_memory(result_block_id, &current_index, &new_value)?;
         }
 
-        let new_slice_val = AcirValue::Array(new_slice);
         let element_type_sizes =
             if super::arrays::array_has_constant_element_size(&slice_typ).is_none() {
                 Some(self.init_element_type_sizes_array(
                     &slice_typ,
                     slice_contents,
-                    Some(&new_slice_val),
+                    Some(&slice),
                     dfg,
                 )?)
             } else {
                 None
             };
 
-        let value_types = new_slice_val.flat_numeric_types();
+        let mut value_types = slice.flat_numeric_types();
+        // We can safely remove the value types based upon the popped elements size =
+        // as we expect the types to be the same for each element.
+        value_types.drain(0..popped_elements_size);
```

Comment:
```suggestion
        value_types.truncate(value_types.len() - popped_elements_size);
```
To avoid shifting in the front after drain?